### PR TITLE
add room metadata field, update method and update signal

### PR DESCRIPTION
**Pull request by lukasIO**
Additions to the protocol to add support for room metadata as proposed in https://github.com/livekit/livekit-server/issues/115#issuecomment-924623911

### CODE REVIEW

**Comment by FeepsDev** [comment]
You may also need to add the UpdateRoomMetadataRequest message to [RTCNodeMessage](https://github.com/livekit/protocol/blob/e3e7ce02322068f1ce25fe9eddf3edb9ea798e2e/livekit_internal.proto#L65)

File: livekit_room.proto [hunk 1]
```diff
@@ -39,6 +39,9 @@ service RoomService {
 
   // Send data over data channel to participants in a room, Requires `roomAdmin`
   rpc SendData(SendDataRequest) returns (SendDataResponse);
+
+  // Update room metadata, will cause updates to be broadcasted to everyone in the room, Requires `roomAdmin`
+  rpc UpdateRoomMetadata (SendDataRequest) returns (SendDataResponse);
```
[the remark this answers is not in the document]
Comment (lukasIO) [reply livekit_room.proto]: thanks! pushed an update

**Comment by lukasIO** [comment]
> 
> 
> You may also need to add the UpdateRoomMetadataRequest message to [RTCNodeMessage](https://github.com/livekit/protocol/blob/e3e7ce02322068f1ce25fe9eddf3edb9ea798e2e/livekit_internal.proto#L65)

thank you! I added it in 43870f0.

> I can do the server implementation if you want ;)
Sure, I'm anyways mostly guessing what the right implementation would be, just have a bit of spare time at the moment to get things moving :)

File: livekit_room.proto [hunk 2]
```diff
@@ -141,3 +144,9 @@ message SendDataRequest {
 message SendDataResponse {
   //
 }
+
+message UpdateRoomRequest {
+  string room = 1;
```
Comment (lukasIO) [inline livekit_room.proto:149]: @FeepsDev I wasn't sure if we'd need an identity field here as well? I left it out for now, but start to think that was a mistake...

File: livekit_internal.proto [hunk 3]
```diff
@@ -74,6 +74,7 @@ message RTCNodeMessage {
     DeleteRoomRequest delete_room = 7;
     UpdateSubscriptionsRequest update_subscriptions = 8;
     SendDataRequest send_data = 9;
+    UpdateRoomRequest update_room = 10;
```
Comment (davidzhao) [inline livekit_internal.proto:77]:
```suggestion
    UpdateRoomMetadataRequest update_room_metadata = 10;
```

File: livekit_room.proto [hunk 4]
```diff
@@ -39,6 +39,9 @@ service RoomService {
 
   // Send data over data channel to participants in a room, Requires `roomAdmin`
   rpc SendData(SendDataRequest) returns (SendDataResponse);
+
+  // Update room metadata, will cause updates to be broadcasted to everyone in the room, Requires `roomAdmin`
+  rpc UpdateRoomMetadata (UpdateRoomRequest) returns (Room);
```
Comment (davidzhao) [inline livekit_room.proto:44]:
```suggestion
  rpc UpdateRoomMetadata (UpdateRoomMetadataRequest) returns (Room);
```

File: livekit_room.proto (hunk 2 above)
Comment (davidzhao) [reply livekit_room.proto:149]: I don't think identity is needed. it is a room-level request.

**Review by davidzhao — COMMENTED** [review]
this looks good! Re: the differences of UpdateRoom vs UpdateRoomMetadata, I think the server API to update should be more focused. Since the use cases are likely changing the room metadata frequently, but not the other attributes, a focused API seems to make more sense.

on the client side, `RoomUpdate` is more appropriate, since we could notify clients of any room updates via the same message. Doing so will reduce the amount of state handling code we'd have to write on the clients.

File: livekit_room.proto (hunk 4 above)
Comment (FeepsDev) [reply livekit_room.proto:44]: it should returns UpdateRoomMetadataResponse

File: livekit_room.proto [hunk 5]
```diff
@@ -141,3 +144,9 @@ message SendDataRequest {
 message SendDataResponse {
   //
 }
+
+message UpdateRoomRequest {
+  string room = 1;
+  // metadata to update. skipping updates if left empty
+  string metadata = 3;
```
Comment (davidzhao) [inline livekit_room.proto:151]:
```suggestion
  string metadata = 2;
```
fields should be sequential, unless we've removed fields.

**Review by davidzhao — COMMENTED** [review]
This is looking good! Just one nit and we are good to merge.

File: livekit_room.proto [hunk 6]
```diff
@@ -39,6 +39,9 @@ service RoomService {
 
   // Send data over data channel to participants in a room, Requires `roomAdmin`
   rpc SendData(SendDataRequest) returns (SendDataResponse);
+
+  // Update room metadata, will cause updates to be broadcasted to everyone in the room, Requires `roomAdmin`
+  rpc UpdateRoomMetadata (UpdateRoomMetadataRequest) returns (Room);
```
Comment (FeepsDev) [inline livekit_room.proto:44]: It should returns UpdateRoomMetadataResponse message

Comment (davidzhao) [reply livekit_room.proto:44]: I think this is actually similar to the other APIs.. where Create/Update/Get returns the object itself without wrapping.

Comment (FeepsDev) [reply livekit_room.proto:44]: ah ok, I thought rather that this API would return nothing

**Comment by davidzhao** [comment]
Tagged as v0.9.4, ready for the server side now.

**Comment by lukasIO** [comment]
great, thank you for the help!

**Pull request merged**